### DISCUSSION TITLE
proposal on introducing binary calculations in Ampersand

### DISCUSSION THREAD

**Discussion by hanjoosten** [AmpersandTarski/Ampersand#1538] [Ideas]
Now we have the REPESENT statement in place, I think it would be nice to introduce calculations. I have been thinking about this for some time now, and I want to share my thoughts on this with you all:
1. Currently, we require that every relation is declared by the user. I suggest to introduce a well defined set of 'special' relation names, like **_gT**_, **_lT**_, **_eQ**_, with pre-defined semantics in Ampersand. These relations can be used in expressions, without the user having to declare them. (declarations by the user with these names might even be forbidden). (note that I do not intent to change the parser)
2. The disambiguator will disregard these relations, somewhat the same as **_V**_ was used. 
3. An extra check (after disambiguation) will be introduced to check the compatibility of the represents of source and target. (at first, I suggest that the represents must be equal)
4. In generating SQL, expressions containing these special relations are treated the way they are currently treated , except for the _EDcD_ constructor. That is handled as _EDcV_, using a binary operator that depends on the semantics of the relationname.

These modifications all seem pretty straitforeward, I suggest to introduce a haskell data type for these pre-defined relations (to guarantee maintainability of it all).  

I think this will allow us to experiment with binary calculations (Haskell type: _a -> a -> Bool_). It is not a solution however for calculations with another haskell type.

Please comment on this proposal (specially @sjcjoosten and @stefjoosten )

**Comment by sjcjoosten** [comment]
There are two ways to implement gt, lt and eq.

Option 1: directly as relations, as you suggest per point 4.
Option 1b: directly as relations, but using the "lt", "gt" etc from PHP. This was once implemented, but only very briefly (stopped working after a while. Was cleaned up later)
Option 2: as a function (UNI/TOT relation): IntPair -> Bool (Haskell type: (a,a) -> Bool)
where IntPair is a relation with "first: IntPair -> Int" and "second: IntPair -> Int" such that first and second form its key.

The last option is trickiest to implement, but would extend to calculations with another type. The first step would be to implement it for "nextInt :: Int -> Int", or "tomorrow:: Date -> Date"

**Comment by hanjoosten** [comment]
In hindsight, this my proposal isn't about calculation at all. It is only about comparison. 

Option 1 just does that. a set of function names is pre-defined, and we implement a function (Haskell type String -> (a,b) -> Bool ). It uses the relations from SQL. 

Option 1b isn't the way to go, for we want to get rid of PHP in the framework (See #197)

Option 2 is an interesting one. It is trickier indeed, and I wonder if it is another solution to the same problem, or a solution to another problem. Functions like these have a different Haskell type. 
In option 2, we need some more restrictions on the relation. I expect it should not only be [UNI,TOT], but even stricter: [UNI,TOT,SUR,INJ]. This way, nextInt~ could be calculated too, when only knowing the target set of that relation. 

Both options have in common that a pre-defined set of relation names should exist. Each of these names should have an easy-to-understand semantics. for every such thing **_x**_, there can be restrictions on the **represents** (not on the Concept name) of source and target of the relation. These restrictions should be checked in the type-checking phase, so the user will get an error message when appropriate. These restrictions should be defined as a function of type (Representation -> Representation -> Bool). 
In other words, I suggest to introduce a Haskell data structure for these pre-defined relations. Then use that data structure to implement the comparison relations in a neat way. I think, relations like 'succ' could be predefined using such datastructure too. Such datastructure ensures maximum help of Haskell, and also assures flexibility, when used well.

**Comment by sjcjoosten** [comment]
OK! So, let's think about this as if there is already a pattern defined with:

```
Concept "Int"
Relation succ :: Int -> Int
Relation ltE :: Int * Int
Relation gE :: Int * Int
Relation maxInt :: Int * Int [PROP]
RULE "Successor increases number" : (I-maxInt);succ |- gE
```

Then the problem becomes: how to reuse a pattern without re-defining and re-exporting all database tables for it (which, in this example, is not possible).

**Comment by hanjoosten** [comment]
I would rather think about this as

Concept Age
Represent Age Type Integer

the pre-defined relations are not typed, nor user-declared. They just can be used. Semantics is in this Haskell data structure.

<more later, have to go now.>

**Comment by hanjoosten** [comment]
The Haskell data structure should contain a list of pre-defined relation names. Each one of them should have a function defined of type Represent ->Represent -> Bool. That function should be used to check if the representations of source and target concepts are valid for that relation. 
The semantics of the relations are either of type  **_a -> b -> Bool**_ or of type **_a -> a**_., so there should be something in the detastructure to reflect that.
Also, the disambiguator should know if there is a restriction on the concept types of the source and target. For the items of type **_a -> b -> Bool**_ there is no restriction. For the relations of type **_a -> a**_, they should be the same. (an endo relation). 
Maybe there is more to know about the pre-defined relations. I suggest to find out by trying to implement this.

**Comment by sjcjoosten** [comment]
I don't see why a->b should be endo.. Haskell function "length :: String -> Integer" is not endo

**Comment by hanjoosten** [comment]
You are right. However, if we try to calculate length, where do we keep the result? There might not be any concept for it. This one adds even more complexity. 

I have created a new branch, where I added [a new module](https://github.com/AmpersandTarski/ampersand/blob/calculations/src/Database/Design/Ampersand/ADL1/PredefinedRelations.hs) for the datastructure I have been talking about. This might give some more insight in what I mean.

**Comment by hanjoosten** [comment]
The idea is, that all semantics of predefined functions should be generalized into the datastructrure. At all places in the current Haskell code where something has to be done with predefined functions, this datastructure must be used to get the nescessary information.

**Comment by Michiel-s** [comment]
@hanjoosten I like your proposal and looking forward to this implementation.

**Comment by hanjoosten** [comment]
I have spoken to Bas, and he has ideas about an even better solution. I hope he will find some time soon for some magic.

[a comment by an automated account is not shown]

**Comment by hanjoosten** [comment]
I saved this issue from becoming stale. It is still relevant.

**Comment by hanjoosten** [comment]
See issue #1534 . This enables the use of several operators as endo-relations
